sys/tests/vm_map.c

```diff
@@ -153,5 +153,107 @@ static int findspace_demo(void) {
   return KTEST_SUCCESS;
 }
 
+static int vm_map_insert_fixed(void) {
+  /* This test mustn't be preempted since PCPU's user-space vm_map will not be
+   * restored while switching back. */
+  SCOPED_NO_PREEMPTION();
+
+  vm_map_t *orig = vm_map_user();
+
+  vm_map_t *umap = vm_map_new();
+  vm_map_activate(umap);
+
+  const vaddr_t addr0 = 0x10000000;
+  const vaddr_t addr1 = 0x10001000;
+  const vaddr_t addr2 = 0x10005000;
+  const vaddr_t addr3 = 0x10009000;
+
+  vm_segment_t *seg;
+  vaddr_t t;
+  int n;
+
+  seg = vm_segment_alloc(NULL, addr0, addr1, VM_PROT_NONE, VM_SEG_PRIVATE);
+  n = vm_map_insert(umap, seg, VM_FIXED);
+  assert(n == 0);
+
+  seg = vm_segment_alloc(NULL, addr2, addr3, VM_PROT_NONE, VM_SEG_PRIVATE);
+  n = vm_map_insert(umap, seg, VM_FIXED);
+  assert(n == 0);
+
+  t = addr1;
+  n = vm_map_findspace(umap, &t, 4 * PAGESIZE);
+  assert(n == 0 && t == addr1);
+
+  t = addr1;
+  n = vm_map_findspace(umap, &t, 5 * PAGESIZE);
+  assert(n == 0 && t == addr3);
+
+  /* Delete first and second mapping and replace them by new one */
+  seg = vm_segment_alloc(NULL, addr1, addr1 + 5 * PAGESIZE, VM_PROT_NONE,
+                         VM_SEG_PRIVATE);
+  n = vm_map_insert(umap, seg, VM_FIXED);
+  assert(n == 0);
+
+  t = addr1;
+  n = vm_map_findspace(umap, &t, 4 * PAGESIZE);
+  assert(n == 0 && t == addr1 + 5 * PAGESIZE);
+
+  vm_map_delete(umap);
+
+  /* Restore original vm_map */
+  vm_map_activate(orig);
+  return KTEST_SUCCESS;
+}
+
+static int vm_map_fixed_bad(void) {
```

Comment:
I am sorry if I mislead you, but I was thinking about the situation when  `vm_map_preparespace_nolock` returns an error (I am not sure if we need this test), but the comment above describes the situation, about which I have been thinking before.

Reply:
Hmm I changed the implementation and now vm_map_preparespace_nolock won't return an error, but vm_map_findspace_nolock will. And this test tests it.